nixos/modules/system/service/systemd/system.nix

```diff
@@ -0,0 +1,90 @@
+{
+  lib,
+  config,
+  options,
+  pkgs,
+  ...
+}:
+
+let
+  inherit (lib)
+    concatMapAttrs
+    mkOption
+    types
+    concatLists
+    mapAttrsToList
+    ;
+
+  portable-lib = import ../portable/lib.nix { inherit lib; };
+
+  dash =
+    before: after:
+    if after == "" then
+      before
+    else if before == "" then
+      after
+    else
+      "${before}-${after}";
+
+  makeUnits =
+    unitType: prefix: service:
+    concatMapAttrs (unitName: unitModule: {
+      "${dash prefix unitName}" =
+        { ... }:
+        {
+          imports = [ unitModule ];
+        };
+    }) service.systemd.${unitType}
+    // concatMapAttrs (
+      subServiceName: subService: makeUnits unitType (dash prefix subServiceName) subService
+    ) service.services;
+in
+{
+  # First half of the magic: mix systemd logic into the otherwise abstract services
+  options = {
+    system.services = mkOption {
+      description = ''
+        A collection of NixOS [modular services](https://nixos.org/manual/nixos/unstable/#modular-services) that are configured as systemd services.
+      '';
+      type = types.attrsOf (
+        types.submoduleWith {
+          class = "service";
+          modules = [
+            ./service.nix
+          ];
```

Comment:
It's awkward that `system.services` is defined in a systemd specific module and the portable services are defined by importing `systemd/service.nix` -> `portable/service.nix`. Modules for other service-managers need to be imported here, as far as I can tell. But I think it's fine for now.

Reply:
A different service manager for `system.services` would imply either
- NixOS offers the choice of a different init system
- or you're using another service manager in addition to systemd

If it's about a different init system, the type of `system.services` could be configured with an option ("dependent type" - trivial in the module system: `type = config.foo;`). Also if it's not about NixOS, e.g. nix-darwin can have its own `system.services` implementation, and I see little opportunity to share code from this file with a nix-darwin provider implementation. (As opposed to the `portable/` files of course)

If it's for an additional service manager, it should have its own option instead, which isn't `system.services`.